envconfig_test.go

```diff
@@ -1596,6 +1596,68 @@ func TestProcessWith(t *testing.T) {
 				"VCR_REMOTE_BUTTON_NAME": "button",
 			}),
 		},
+
+		// No preemptive init
+		{
+			name: "nopreemptiveinit/no_init_when_sub_fields_set",
+			input: &struct {
+				Sub *struct {
+					Field string `env:"FIELD"`
+				} `env:",nopreemptiveinit"`
+			}{},
+			exp: &struct {
+				Sub *struct {
+					Field string `env:"FIELD"`
+				} `env:",nopreemptiveinit"`
+			}{},
+			// 'Sub' struct ptr shouldn't be initiated
+			// because the 'Field' is not set in the lookuper.
+			lookuper: MapLookuper(map[string]string{}),
+		},
+		{
+			name: "nopreemptiveinit/init_when_sub_fields_set",
+			input: &struct {
+				Sub *struct {
+					Field string `env:"FIELD"`
+				} `env:",nopreemptiveinit"`
+			}{},
+			exp: &struct {
+				Sub *struct {
+					Field string `env:"FIELD"`
+				} `env:",nopreemptiveinit"`
+			}{
+				Sub: &struct {
+					Field string `env:"FIELD"`
+				}{
+					Field: "banana",
+				},
+			},
+			// 'Sub' struct ptr should be initiated
+			// because the 'Field' is set in the lookuper.
+			lookuper: MapLookuper(map[string]string{
+				"FIELD": "banana",
+			}),
+		},
```

Comment:
Let's add a test for what happens to a non-pointer set with noinit.

```go
struct {
  Sub struct { // non-ptr
    Field string `env:"FIELD"`
  }
}
```

This is mostly to ensure we don't panic on that case.

For discussion: what should we do if someone specifies noinit on a non-pointer field? Should that be a runtime error?

Reply:
I think it's better to ignore the tag if it's a non struct ptr field, like "not applicable so no effect". To make it more obvious, we could name it as `noptrinit`?

Reply:
I also update the readme saying apply the tag on non-struct-pointer fields won't have any effect.

Reply:
I think `noinit` is the right name. I'm wondering if we should throw an error though. We already throw errors for other "impossible" configurations (e.g. specifying both _default_ and _required_)